### DISCUSSION TITLE
.on_conflict or upsert

### DISCUSSION THREAD

**Discussion by TheKnudsen** [piccolo-orm/piccolo#248] [Q&A] [answered]
What is the equivalent in Piccolo to .on_conflict or upsert? 
If not, is there a way to capture the id from the violating record via an Exception?

**Comment by dantownsend** [comment]
@Thaiwegian There isn't an API for upsert yet, but I'll create a new issue for it.

For now, here's a workaround:

```python
class Person(Table):
    name = Varchar()
    age = Integer()


values = {Person.age: 33}
person = Person.objects().get_or_create(Person.name == 'Bob', values).run_sync()
if not person._was_created:
    Person.update(defaults).where(Person.name == 'Bob').run_sync()
```

I hope that helps.

**Reply by dantownsend** [reply]
Created issue: https://github.com/piccolo-orm/piccolo/issues/252

**Reply by TheKnudsen** [reply]
Thanks. This is good for user registration and things like that, but for bulk inserts, probably not. I will try to use Table.raw() for now, but it's not a showstopper.

**Comment by dantownsend** [comment, marked as answer]
Upsert is supported now: https://piccolo-orm.readthedocs.io/en/latest/piccolo/query_clauses/on_conflict.html
